core/finality-grandpa/src/lib.rs

```diff
@@ -241,6 +264,158 @@ impl Stream for NetworkStream {
 	}
 }
 
+struct TopicTracker {
+	min_live_round: u64,
+	max_round: u64,
+	set_id: u64,
+}
+
+struct GossipValidator<Block: BlockT> {
+	rounds: parking_lot::RwLock<TopicTracker>,
+	_marker: ::std::marker::PhantomData<Block>,
+}
+
+impl<Block: BlockT> GossipValidator<Block> {
+	fn new() -> GossipValidator<Block> {
+		GossipValidator {
+			rounds: parking_lot::RwLock::new(TopicTracker {
+				min_live_round: 0,
+				max_round: 0,
+				set_id: 0,
+			}),
+			_marker: Default::default(),
+		}
+	}
+
+	fn note_round(&self, round: u64, set_id: u64) {
+		let mut rounds = self.rounds.write();
+		if set_id > rounds.set_id {
+			rounds.set_id = set_id;
+			rounds.max_round = 0;
+			rounds.min_live_round = 0;
+		}
+		rounds.max_round = rounds.max_round.max(round);
+	}
+
+	fn note_set(&self, _set_id: u64) {
+	}
+
+	fn drop_round(&self, round: u64, set_id: u64) {
+		let mut rounds = self.rounds.write();
+		if set_id == rounds.set_id && round >= rounds.min_live_round {
+			rounds.min_live_round = round + 1;
+		}
+	}
+
+	fn drop_set(&self, _set_id: u64) {
+	}
+
+	fn is_expired(&self, round: u64, set_id: u64) -> bool {
+		let rounds = self.rounds.read();
+		if set_id < rounds.set_id {
+			trace!(target: "afg", "Expired: Message with expired set_id {} (ours {})", set_id, rounds.set_id);
+			return true;
+		} else if set_id == rounds.set_id + 1 {
+			// allow a few first rounds of future set.
+			if round > MESSAGE_ROUND_TOLERANCE {
```

Comment:
what's the strategy for when we catch up? if messages are dropped consensus will halt.

Reply:
After discussing it with @rphmeier and @andresilva here's what the issue here is:
Grandpa starts round N+1 when he observes round commit for round N and requires all messages for round N+1 to complete it. Currently there's no guarantee that commit message will be the first in the set of messages for N+1, so some messages might be skipped if current round is far behind N. 

There are a few solutions:
1. Request missing messages from peers when we observe commit N and it is more than a few rounds in the future. Would require substantial network protocol changes.

2. GRANDPA begins round N+2 when he observes commit for N. This would guarantee that all messages for N+2 are accepted as they arrive only after commit for N. Appears to be involved change in the grandpa crate. 

3. Future messages are kept around with time based expiration. Will be implemented in this PR.

Additional DOS protection scheme will be implemented in the future.

Reply:
1 seems like a sensible thing to implement longer-term.